modules/tensor_mechanics/src/materials/ADLAROMCreepStressUpdate.C

```diff
@@ -0,0 +1,535 @@
+  std::vector<ADReal> drom_outputs(_num_outputs);
+  std::vector<ADReal> dinput_value_increments(_num_outputs);
+
+  buildPolynomials(rom_inputs, drom_inputs, polynomial_inputs, dpolynomial_inputs);
+  computeValues(_coefs, polynomial_inputs, dpolynomial_inputs, rom_outputs, drom_outputs);
+  convertOutput(dt,
+                old_input_values,
+                rom_outputs,
+                drom_outputs,
+                input_value_increments,
+                dinput_value_increments);
+
+  mobile_dislocation_increment = input_value_increments[0];
+  immobile_dislocation_increment = input_value_increments[1];
+  rom_effective_strain = input_value_increments[2];
+
+  rom_effective_strain_derivative = dinput_value_increments[2];
+}
+
+template <ComputeStage compute_stage>
+void
+ADLAROMCreepStressUpdate<compute_stage>::checkInputWindows(std::vector<ADReal> & input)
+{
+  if (compute_stage != RESIDUAL)
+    return;
+
+  _failed[_qp] = 0.0;
+  for (unsigned int i = 0; i < _num_outputs; ++i)
+  {
+    for (unsigned int j = 0; j < _num_inputs; ++j)
+    {
+      Real high_limit = _input_limits[j][1] * _window;
+      Real low_limit = _input_limits[j][0] * (2.0 - _window);
+      if (j == _stress_index && _extrapolate_stress)
+        low_limit = 0.0;
+      if (input[j] < low_limit || input[j] > high_limit)
+      {
+        _failed[_qp] += (j + 1) * (j + 1);
+        if (_window_failure == WindowFailure::WARN)
+          mooseWarning("In ",
+                       _name,
+                       ": input parameter number input=",
+                       j,
+                       " output=",
+                       i,
+                       " with value (",
+                       input[j],
+                       ") is out of range (",
+                       _input_limits[j][0],
+                       " - ",
+                       _input_limits[j][1],
+                       "), window (",
+                       _window,
+                       ")");
+        else if (_window_failure == WindowFailure::ERROR)
+          mooseError("In ",
+                     _name,
+                     ": input parameter number input=",
+                     j,
+                     " output=",
+                     i,
+                     " with value (",
+                     input[j],
+                     ") is out of range (",
+                     _input_limits[j][0],
+                     " - ",
+                     _input_limits[j][1],
+                     "), window (",
+                     _window,
+                     ")");
+      }
+    }
+  }
+}
+
+template <ComputeStage compute_stage>
+void
+ADLAROMCreepStressUpdate<compute_stage>::convertInput(const std::vector<ADReal> & input,
+                                                      std::vector<std::vector<ADReal>> & converted,
+                                                      std::vector<std::vector<ADReal>> & dconverted)
```

Comment:
I like the reuse of the same methods to hold both the original functions and the derivatives